ir/visitor.h

```diff
@@ -479,6 +498,121 @@ class ControlFlowVisitor : public virtual Visitor {
     };
 };
 
+class SplitFlowVisit_base {
+ protected:
+    Visitor                     &v;
+    SplitFlowVisit_base         *prev;
+    std::vector<Visitor *>      visitors;
+    int                         visit_next, start_index;
+    friend ControlFlowVisitor;
+
+    explicit SplitFlowVisit_base(Visitor &v) : v(v), visit_next(0) {
+        prev = v.split_link;
+        v.split_link = this; }
+    ~SplitFlowVisit_base() { v.split_link = prev; }
+
+ public:
+    bool finished() { return size_t(visit_next) >= visitors.size(); }
+    virtual void do_visit() = 0;
+    virtual void run_visit() {
+        auto *ctxt = v.getChildContext();
+        start_index = ctxt ? ctxt->child_index : 0;
+        while (!finished())
+            do_visit();
+        bool first = true;
+        for (auto *cl : visitors) {
+            if (first)
+                first = false;
+            else
+                v.flow_merge(*cl); } }
+};
+
+template<class N> class SplitFlowVisit : public SplitFlowVisit_base {
+    std::vector<const N **> nodes;
+    std::vector<const N *const *> const_nodes;
+
+ public:
+    explicit SplitFlowVisit(Visitor &v) : SplitFlowVisit_base(v) {}
+    void addNode(const N *&node) {
+        BUG_CHECK(const_nodes.empty(), "Mixing const and non-const in SplitFlowVisit");
```

Comment:
maybe these 2 common checks could be moved to a "validate" method?

Reply:
The checks are subtly different (dealing with `const_nodes` and `nodes` in the const and non-const versions of the functions.)